packages/core/src/awsService/ec2/remoteSessionManager.ts

```diff
@@ -0,0 +1,41 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { EC2, SSM } from 'aws-sdk'
+import { SsmClient } from '../../shared/clients/ssmClient'
+
+export class Ec2RemoteSessionManager {
+    private activeSessions: Map<EC2.InstanceId, SSM.SessionId>
+
+    public constructor(
+        readonly regionCode: string,
+        protected ssmClient: SsmClient
+    ) {
+        this.activeSessions = new Map<EC2.InstanceId, SSM.SessionId>()
+    }
+
+    public async addSession(instanceId: EC2.InstanceId, sessionId: SSM.SessionId): Promise<void> {
+        if (this.isConnectedTo(instanceId)) {
+            const existingSessionId = this.activeSessions.get(instanceId)!
+            await this.ssmClient.terminateSessionFromId(existingSessionId)
+            this.activeSessions.set(instanceId, sessionId)
+        } else {
+            this.activeSessions.set(instanceId, sessionId)
+        }
+    }
+
+    private async disconnectEnv(instanceId: EC2.InstanceId): Promise<void> {
+        await this.ssmClient.terminateSessionFromId(this.activeSessions.get(instanceId)!)
+        this.activeSessions.delete(instanceId)
+    }
+
+    public async closeConnections(): Promise<void> {
```

Comment:
maybe calling this dispose() is a useful signal (and the class could implement Disposable)
```suggestion
    public async dispose(): Promise<void> {
```